ophyd/status.py

```diff
@@ -127,25 +170,74 @@ def _settled(self):
         """Hook for when status has completed and settled"""
         pass
 
-    def _settle_then_run_callbacks(self, success=True):
+    def _settle_then_run_callbacks(self):
+        """
+        Sleep for the settle_time, set the Event, run the callbacks.
+        """
         # wait until the settling time is done to mark completion
         if self.settle_time > 0.0:
             time.sleep(self.settle_time)
 
         with self._lock:
-            if self.done:
+            if self._event.is_set():
                 # We timed out while waiting for the settle time.
                 return
-            self.success = success
-            self._done = True
+            self._event.set()
             self._settled()
 
             for cb in self._callbacks:
                 cb(self)
             self._callbacks.clear()
 
+    def set_exception(self, exc):
+        """
+        Mark as finished but failed with the given Exception.
+
+        This method should generally not be caled by the *recipient* of this
+        Status object, but only by the object that created and returned it.
+
+        Parameters
+        ----------
+        exc: Exception
+        """
+        # Since we rely on this being raise-able later, check proactively to
+        # avoid potentially very confusing failures.
+        if not isinstance(exc, Exception):
+            raise ValueError(f"Expected an Exception, got {exc!r}")
+        with self._lock:
+            if self._event.is_set():
+                raise RuntimeError(
+                    f"set_exception was called more than once on {self!r}")
+            self._exception = exc
+            self._event.set()
+
+    def set_finished(self):
+        """
+        Mark as finished successfully.
+
+        This method should generally not be caled by the *recipient* of this
+        Status object, but only by the object that created and returned it.
+        """
+        if self.done:
+            # This is fast path. We do a proper check inside a lock in the call
+            # to self._settle_then_run_callbacks.
+            return
+
+        if self.settle_time > 0:
+            # delay gratification until the settle time is up
+            self._settle_thread = threading.Thread(
```

Comment:
Consider replacing with https://docs.python.org/3/library/threading.html#threading.Timer and removing the `time.sleep`